[Fix #11024] Add `require_always` option to `Style/EndlessMethod`

koic opened the pull request:
Fixes #11024.

This PR adds `require_always` option to `Style/EndlessMethod` that all endless method definitions are required.

```ruby
# EnforcedStyle: require_always

# bad
def my_method
  x
end

# bad
def my_method
  x.foo
   .bar
   .baz
end

# good
def my_method = x

# good
def my_method = x.foo
                 .bar
                 .baz
```

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

bbatsov commented:
We'll have to think about the naming (or at least improve the high-level description for the new style), though, as now I find it to be somewhat confusing.

koic commented:
I've updated to `EnforcedStyle: require_endless_method`.

The `require_endless_method` style enforces endless method definitions for a single statement (not a single line) containing a method chain with a line break.

```ruby
# EnforcedStyle: require_endless_method

# bad
def my_method
  x
end

# bad
def my_method
  x.foo
   .bar
   .baz
end

# good
def my_method = x

# bad
def my_method = x.foo
                 .bar
                 .baz
```

Possibly a bit different than what #11024 wants, but it's simpler in name and behavior and always enforces `EnforcedStyle: require_endless_method`. cc @jonas054 

I also thought of `EnforcedStyle: single_line_endless_method` which allows the following case.

```ruby
# EnforcedStyle: single_line_endless_method

# bad
def my_method
  x
end

# bad
def my_method = x.foo
                 .bar
                 .baz

# good
def my_method = x

# good
def my_method
  x.foo
   .bar
   .baz
end
```

But its name and behavior are a little more complicated than `EnforcedStyle: require_endless_method`.

So, I've updated this PR using `EnforcedStyle: require_endless_method`. How about?

bbatsov commented:
> require_endless_method

Or perhaps `always_require`?  I'll have to think about the naming here a bit more and how this will tie with `always_for_single_line` or something like this.

koic commented:
> Or perhaps `always_require`?

This name represents the current implementation that this PR adds. OHTH, there is already an option called `allow_always`, I updated it to `require_always` instead.

```diff
       # * allow_single_line (default) - only single line endless method definitions are allowed.
       # * allow_always - all endless method definitions are allowed.
       # * disallow - all endless method definitions are disallowed.
+      # * require_always - all endless method definitions are required.
```

jonas054 commented on lib/rubocop/cop/style/endless_method.rb:
```diff
@@ -5,8 +5,11 @@ module Cop
+      # or disallow endless methods.
+      #
+      # `require_always` style enforces endless method definitions for a single statement
+      # (not a single line) containing a method chain with a line break.
```
I think the second line of this paragraph is a bit confusing, because the implementation of this style doesn't check whether there are any line breaks, and doesn't care if it's a method chain och just a numeric literal for example. So:

`require_always` style enforces endless method definitions for single statement methods.

koic replied:
I like the proposed text, it's simple! 🙏  I've update it!

jonas054 commented on lib/rubocop/cop/style/endless_method.rb:
```diff
@@ -5,8 +5,11 @@ module Cop
-      # It can enforce either the use of endless methods definitions
-      # for single-lined method bodies, or disallow endless methods.
+      # It can enforce either the use of endless methods definitions for single-lined method bodies,
+      # or disallow endless methods.
```
It's difficult to describe all the different styles now in a single compact paragraph, since it's a mixture of enforcing and allowing, and sometimes for single lines and sometimes not. Here's my suggestion:

It can enforce endless method definitions whenever possible. It can also disallow multiline endless method definitions or all endless definitions.

jonas054 commented on lib/rubocop/cop/style/endless_method.rb:
```diff
@@ -15,36 +18,94 @@ module Style
       # * allow_single_line (default) - only single line endless method definitions are allowed.
       # * allow_always - all endless method definitions are allowed.
```
By the way, is there any difference between `EnforcedStyle: allow_always` and `Enabled: false`?

koic replied:
Ah, indeed it is!
It's possible that the option will be deprecated, but the incompatibility is annoying.

jonas054 commented on lib/rubocop/cop/style/endless_method.rb:
```diff
@@ -76,6 +141,14 @@ def handle_allow_style(node)
+        def handle_require_always_style(node)
+          return if node.endless? || !node.body || node.body.begin_type? || node.body.kwbegin_type?
+
+          add_offense(node, message: MSG_REQUIRE_ALWAYS) do |corrector|
```
I see a possible problem if we report and autocorrect methods into endless methods that are too long for a single line. On the other hand, there's autocorrection in `Layout/LineLength` too, so some of these cases will be solved by the autocorrection loop.

koic replied:
I wasn't aware of this problem. I will take a look too.

kalivara replied:
```ruby
def handle_require_always_style(node)
  return if node.endless? || !node.body || node.body.begin_type? || node.body.kwbegin_type?

  line_length = node.body.source.length + definition.length
  return if line_length > config_store.line_length # or however this works

  add_offense(node, message: MSG_REQUIRE_ALWAYS) do |corrector|
    correct_to_endless_method(corrector, node)
  end
end

def definition(node) = "def #{node.method_name}#{arguments(node)} = "

def correct_to_endless_method(corrector, node)
  replacement = "#{definition}#{node.body.source}"
  corrector.replace(node, replacement)
end
```

bbatsov commented:
After thinking about this a bit more - without something like `require_single_line` I think that the cop will remain to be pretty weird. I think the use cases for `require_always` will be relatively limited in comparison.

metalelf0 commented:
Is there any update on this PR? Can I help somehow to get this merged?

koic commented:
I was late retaking this. I will retake in a few days.

bbatsov commented:
@koic One more forgotten PR. :-)

smcabrera commented:
> I was late retaking this. I will retake in a few days.

Hey @koic any updates on this? Anything I can do to help?

ttanimichi commented:
I think one line method definition should be replaced to endless method definition.

```ruby
# bad
def edit; end

# good
def edit = nil
```

Abdullah-l commented:
This PR is stale for a year, what can we do to get it merged?

lvonk commented:
It would be great if this came available... 🙏🏼 any help needed?

jtannas commented:
Probably a task for a follow-up PR, but `require_single_line` would be another nice option to have.
Please and thank you :bow: 
```
      # @example EnforcedStyle: require_single_line
      #   # bad
      #   def my_method
      #     x
      #   end
      #
      #   # bad
      #   def my_method = x.foo
      #                    .bar
      #                    .baz
      #
      #   # good
      #   def my_method = x
      #
      #   # good
      #   def my_method
      #     x.foo
      #      .bar
      #      .baz
      #   end
```

jtannas commented:
@koic I took a crack at building on top of your work to implement `require_single_line` if you're interested
https://github.com/koic/rubocop/pull/4

jtannas commented:
@koic It seems you're busy with other work.
My apologies if this is premature, but do you mind if I take over this task to get it over the finish line?
Please and thank you :bow:

koic commented:
Closing this as it has been addressed in #13874.
